File: backend/old/src/services/mainConfigService.ts
/**
 * 麦麦主程序配置服务
 * 提供对bot_config.toml文件的读取和写入操作
 */

import { logger } from '../core/logger';
import { pathCacheManager } from '../core/pathCacheManager';
import { 
  MainConfigData, 
  ConfigUpdateData, 
  ConfigFileInfo, 
  ConfigServiceOptions 
} from '../types/config';
import { ValidationError, NotFoundError, InternalServerError } from '../middleware/errorHandler';
import { stringifyConfigToCleanToml } from '../utils/tomlHelpers';
import * as TOML from '@iarna/toml';
import * as fs from 'fs';
import * as path from 'path';export class MainConfigService {
  private static readonly CONFIG_FILE_NAME = 'bot_config.toml';
  private static readonly CONFIG_DIR = 'config';
  private static readonly DEFAULT_OPTIONS: ConfigServiceOptions = {
    encoding: 'utf-8',
    backup: true,
    validate: true
  };

  /**
   * 获取配置文件的完整路径
   */
  private static getConfigPath(): string {
    const mainRoot = pathCacheManager.getMainRoot();
    if (!mainRoot) {
      throw new NotFoundError('麦麦主程序根目录未设置，请先设置根目录缓存');
    }

    return path.join(mainRoot, this.CONFIG_DIR, this.CONFIG_FILE_NAME);
  }

  /**
   * 检查配置文件信息
   */
  public static async getConfigFileInfo(): Promise<ConfigFileInfo> {
    try {
      const configPath = this.getConfigPath();
      
      let stats: fs.Stats | null = null;
      let exists = false;
      
      try {
        stats = await fs.promises.stat(configPath);
        exists = true;
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code !== 'ENOENT') {
          throw error;
        }
      }

      const info: ConfigFileInfo = {
        path: configPath,
        exists,
        readable: false,
        writable: false,
        size: 0,
        lastModified: new Date(0)
      };

      if (exists && stats) {
        info.size = stats.size;
        info.lastModified = stats.mtime;

        // 检查读写权限
        try {
          await fs.promises.access(configPath, fs.constants.R_OK);
          info.readable = true;
        } catch {
          info.readable = false;
        }

        try {
          await fs.promises.access(configPath, fs.constants.W_OK);
          info.writable = true;
        } catch {
          info.writable = false;
        }
      }

      return info;
    } catch (error) {
      logger.error('获取配置文件信息失败:', error);
      throw new InternalServerError('获取配置文件信息失败');
    }
  }

  /**
   * 读取主程序配置文件
   */
  public static async getConfig(options: Partial<ConfigServiceOptions> = {}): Promise<MainConfigData> {
    try {
      const opts = { ...this.DEFAULT_OPTIONS, ...options };
      const configPath = this.getConfigPath();

      // 检查文件是否存在
      const fileInfo = await this.getConfigFileInfo();
      if (!fileInfo.exists) {
        throw new NotFoundError(`主程序配置文件不存在: ${configPath}`);
      }

      if (!fileInfo.readable) {
        throw new InternalServerError(`主程序配置文件不可读: ${configPath}`);
      }

      // 读取文件内容
      let fileContent: string;
      try {
        const content = await fs.promises.readFile(configPath, { encoding: opts.encoding });
        fileContent = typeof content === 'string' ? content : content.toString();
      } catch (error) {
        logger.error('读取主程序配置文件失败:', error);
        throw new InternalServerError(`读取主程序配置文件失败: ${error instanceof Error ? error.message : '未知错误'}`);
      }

      // 解析TOML内容
      let configData: MainConfigData;
      try {
        configData = TOML.parse(fileContent) as MainConfigData;
      } catch (error) {
        logger.error('解析主程序配置文件失败:', error);
        throw new ValidationError(`主程序配置文件格式错误: ${error instanceof Error ? error.message : 'TOML解析失败'}`);
      }

      logger.info('主程序配置读取成功');
      return configData;

    } catch (error) {
      logger.error('获取主程序配置失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError || error instanceof InternalServerError
        ? error 
        : new InternalServerError('获取主程序配置失败');
    }
  }

  /**
   * 更新主程序配置文件
   */
  public static async updateConfig(
    updateData: ConfigUpdateData, 
    options: Partial<ConfigServiceOptions> = {}
  ): Promise<void> {
    try {
      const opts = { ...this.DEFAULT_OPTIONS, ...options };
      const configPath = this.getConfigPath();

      if (!updateData || Object.keys(updateData).length === 0) {
        throw new ValidationError('更新数据不能为空');
      }

      // 验证更新数据格式
      if (opts.validate) {
        this.validateConfigData(updateData);
      }

      // 检查文件状态
      const fileInfo = await this.getConfigFileInfo();
      
      let currentConfig: MainConfigData = {};
      
      // 如果文件存在，先读取当前配置
      if (fileInfo.exists) {
        if (!fileInfo.readable) {
          throw new InternalServerError(`主程序配置文件不可读: ${configPath}`);
        }
        if (!fileInfo.writable) {
          throw new InternalServerError(`主程序配置文件不可写: ${configPath}`);
        }
        
        currentConfig = await this.getConfig({ validate: false });
      } else {
        // 确保目录存在
        const configDir = path.dirname(configPath);
        await fs.promises.mkdir(configDir, { recursive: true });
      }

      // 合并配置数据
      const mergedConfig = this.mergeConfig(currentConfig, updateData);

      // 创建备份（如果文件存在且启用备份）
      if (fileInfo.exists && opts.backup) {
        await this.createBackup(configPath);
      }

      // 转换为TOML格式并写入文件
      try {
        const tomlContent = stringifyConfigToCleanToml(mergedConfig, TOML);
        await fs.promises.writeFile(configPath, tomlContent, { encoding: opts.encoding });
      } catch (error) {
        logger.error('写入主程序配置文件失败:', error);
        throw new InternalServerError(`写入主程序配置文件失败: ${error instanceof Error ? error.message : '未知错误'}`);
      }

      logger.info('主程序配置更新成功');

    } catch (error) {
      logger.error('更新主程序配置失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError || error instanceof InternalServerError
        ? error 
        : new InternalServerError('更新主程序配置失败');
    }
  }

  /**
   * 验证配置数据格式
   */
  private static validateConfigData(data: ConfigUpdateData): void {
    // 基础类型检查
    if (typeof data !== 'object' || data === null || Array.isArray(data)) {
      throw new ValidationError('配置数据必须是一个对象');
    }

    // 递归检查嵌套对象，确保数据结构合理
    const validateObject = (obj: any, path = ''): void => {
      for (const [key, value] of Object.entries(obj)) {
        const currentPath = path ? `${path}.${key}` : key;
        
        // 检查键名
        if (typeof key !== 'string' || key.trim() === '') {
          throw new ValidationError(`配置项键名无效: ${currentPath}`);
        }

        // 检查值类型（TOML支持的类型）
        const valueType = typeof value;
        const isValidType = [
          'string', 'number', 'boolean'
        ].includes(valueType) || 
        value === null || 
        Array.isArray(value) || 
        (valueType === 'object' && value !== null);

        if (!isValidType) {
          throw new ValidationError(`配置项类型不支持: ${currentPath} (${valueType})`);
        }

        // 递归验证嵌套对象
        if (valueType === 'object' && value !== null && !Array.isArray(value)) {
          validateObject(value, currentPath);
        }

        // 验证数组元素类型一致性
        if (Array.isArray(value) && value.length > 0) {
          const firstType = typeof value[0];
          const allSameType = value.every(item => typeof item === firstType);
          if (!allSameType) {
            throw new ValidationError(`数组元素类型不一致: ${currentPath}`);
          }
        }
      }
    };

    validateObject(data);
  }

  /**
   * 合并配置数据（深度合并）
   */
  private static mergeConfig(current: MainConfigData, update: ConfigUpdateData): MainConfigData {
    const merged = { ...current };

    for (const [key, value] of Object.entries(update)) {
      if (value !== undefined) {
        if (
          typeof value === 'object' && 
          value !== null && 
          !Array.isArray(value) &&
          typeof merged[key] === 'object' && 
          merged[key] !== null && 
          !Array.isArray(merged[key])
        ) {
          // 递归合并嵌套对象
          merged[key] = this.mergeConfig(merged[key], value);
        } else {
          // 直接替换值
          merged[key] = value;
        }
      }
    }

    return merged;
  }

  /**
   * 创建配置文件备份
   */
  private static async createBackup(configPath: string): Promise<void> {
    try {
      const mainRoot = pathCacheManager.getMainRoot();
      if (!mainRoot) {
        throw new Error('无法获取麦麦根目录路径');
      }

      // 创建备份目录结构：LauncherConfigBak/main/
      const backupDir = path.join(mainRoot,'config', 'LauncherConfigBak', 'main');
      await fs.promises.mkdir(backupDir, { recursive: true });

      // 生成带时间戳的备份文件名
      const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
      const configFileName = path.basename(configPath);
      const backupFileName = `${configFileName}.backup.${timestamp}`;
      const backupPath = path.join(backupDir, backupFileName);

      // 复制配置文件到备份目录
      await fs.promises.copyFile(configPath, backupPath);
      logger.info(`主程序配置备份已创建: ${backupPath}`);

    } catch (error) {
      logger.warn('创建配置文件备份失败:', error);
      // 备份失败不应该阻止配置更新，只记录警告
    }
  }
}

File: backend/src/services/main_config_service.py
"""
主程序配置服务
Main Configuration Service
提供对bot_config.toml文件的读取和写入操作
"""

import os
import shutil
from datetime import datetime
from typing import Optional, Dict, Any

import toml
import tomlkit

from core.logger import logger
from core.path_cache_manager import path_cache_manager
from models.config import (
    MainConfigData,
    ConfigUpdateData,
    ConfigFileInfo,
    ConfigServiceOptions
)
from utils.toml_helpers import stringify_config_to_clean_toml


class MainConfigService:
    """主程序配置服务类"""
    
    CONFIG_FILE_NAME = 'bot_config.toml'
    CONFIG_DIR = 'config'
    
    @classmethod
    def _get_config_path(cls) -> str:
        """获取配置文件的完整路径"""
        main_root = path_cache_manager.get_main_root()
        if not main_root:
            raise FileNotFoundError('麦麦主程序根目录未设置，请先设置根目录缓存')
        
        return os.path.join(main_root, cls.CONFIG_DIR, cls.CONFIG_FILE_NAME)
    
    @classmethod
    async def get_config_file_info(cls) -> ConfigFileInfo:
        """检查配置文件信息"""
        try:
            config_path = cls._get_config_path()
            
            exists = os.path.exists(config_path)
            readable = False
            writable = False
            size = 0
            last_modified = datetime.now().isoformat()
            
            if exists:
                stat_info = os.stat(config_path)
                size = stat_info.st_size
                last_modified = datetime.fromtimestamp(stat_info.st_mtime).isoformat()
                
                # 检查读写权限
                readable = os.access(config_path, os.R_OK)
                writable = os.access(config_path, os.W_OK)
            
            return ConfigFileInfo(
                path=config_path,
                exists=exists,
                readable=readable,
                writable=writable,
                size=size,
                last_modified=last_modified
            )
            
        except Exception as error:
            logger.error(f'获取配置文件信息失败: {error}')
            raise RuntimeError('获取配置文件信息失败')
    
    @classmethod
    async def get_config(cls, options: Optional[ConfigServiceOptions] = None) -> MainConfigData:
        """读取主程序配置文件"""
        try:
            opts = options or ConfigServiceOptions()
            config_path = cls._get_config_path()
            
            # 检查文件是否存在
            file_info = await cls.get_config_file_info()
            if not file_info.exists:
                raise FileNotFoundError(f'主程序配置文件不存在: {config_path}')
            
            if not file_info.readable:
                raise PermissionError(f'主程序配置文件不可读: {config_path}')
            
            # 读取文件内容
            try:
                with open(config_path, 'r', encoding=opts.encoding) as f:
                    file_content = f.read()
            except Exception as error:
                logger.error(f'读取主程序配置文件失败: {error}')
                raise RuntimeError(f'读取主程序配置文件失败: {error}')
            
            # 解析TOML内容
            try:
                config_data = toml.loads(file_content)
            except Exception as error:
                logger.error(f'解析主程序配置文件失败: {error}')
                raise ValueError(f'主程序配置文件格式错误: {error}')
            
            logger.info('主程序配置读取成功')
            return MainConfigData(**config_data)
            
        except Exception as error:
            logger.error(f'获取主程序配置失败: {error}')
            raise error
    
    @classmethod
    async def update_config(
        cls,
        update_data: ConfigUpdateData,
        options: Optional[ConfigServiceOptions] = None
    ) -> None:
        """更新主程序配置文件"""
        try:
            opts = options or ConfigServiceOptions()
            config_path = cls._get_config_path()
            
            if not update_data:
                raise ValueError('更新数据不能为空')
            
            update_dict = update_data.model_dump()
            if not update_dict:
                raise ValueError('更新数据不能为空')
            
            # 验证更新数据格式
            if opts.validate:
                cls._validate_config_data(update_dict)
            
            # 检查文件状态
            file_info = await cls.get_config_file_info()
            
            current_config = {}
            
            # 如果文件存在，先读取当前配置
            if file_info.exists:
                if not file_info.readable:
                    raise PermissionError(f'主程序配置文件不可读: {config_path}')
                if not file_info.writable:
                    raise PermissionError(f'主程序配置文件不可写: {config_path}')
                
                current_config_data = await cls.get_config(ConfigServiceOptions(validate=False))
                current_config = current_config_data.model_dump()
            else:
                # 确保目录存在
                config_dir = os.path.dirname(config_path)
                os.makedirs(config_dir, exist_ok=True)
            
            # 合并配置数据
            merged_config = cls._merge_config(current_config, update_dict)
            
            # 创建备份（如果文件存在且启用备份）
            if file_info.exists and opts.backup:
                await cls._create_backup(config_path)
            
            # 转换为TOML格式并写入文件
            try:
                if file_info.exists:
                    # 如果文件存在，使用tomlkit保留注释和格式
                    with open(config_path, 'r', encoding=opts.encoding) as f:
                        original_content = f.read()
                    
                    # 使用tomlkit解析原始内容
                    doc = tomlkit.parse(original_content)
                    
                    # 更新配置值
                    cls._update_tomlkit_doc(doc, merged_config)
                    
                    # 写入更新后的内容
                    with open(config_path, 'w', encoding=opts.encoding) as f:
                        f.write(tomlkit.dumps(doc))
                else:
                    # 如果是新文件，使用标准方法
                    toml_content = stringify_config_to_clean_toml(merged_config, toml)
                    with open(config_path, 'w', encoding=opts.encoding) as f:
                        f.write(toml_content)
            except Exception as error:
                logger.error(f'写入主程序配置文件失败: {error}')
                raise RuntimeError(f'写入主程序配置文件失败: {error}')
            
            logger.info('主程序配置更新成功')
            
        except Exception as error:
            logger.error(f'更新主程序配置失败: {error}')
            raise error
    
    @classmethod
    def _validate_config_data(cls, data: Dict[str, Any]) -> None:
        """验证配置数据格式"""
        if not isinstance(data, dict):
            raise ValueError('配置数据必须是一个对象')
        
        # 递归检查嵌套对象，确保数据结构合理
        def validate_object(obj: Any, path: str = '') -> None:
            if isinstance(obj, dict):
                for key, value in obj.items():
                    current_path = f'{path}.{key}' if path else key
                    
                    # 检查键名
                    if not isinstance(key, str) or not key.strip():
                        raise ValueError(f'配置项键名无效: {current_path}')
                    
                    # 检查值类型（TOML支持的类型）
                    if not cls._is_valid_toml_type(value):
                        raise ValueError(f'配置项类型不支持: {current_path} ({type(value).__name__})')
                    
                    # 递归验证嵌套对象
                    if isinstance(value, (dict, list)):
                        validate_object(value, current_path)
            
            elif isinstance(obj, list):
                # 验证数组元素类型一致性
                if obj and len(obj) > 0:
                    first_type = type(obj[0])
                    if not all(isinstance(item, first_type) for item in obj):
                        raise ValueError(f'数组元素类型不一致: {path}')
                    
                    # 递归验证数组元素
                    for i, item in enumerate(obj):
                        validate_object(item, f'{path}[{i}]')
        
        validate_object(data)
    
    @classmethod
    def _is_valid_toml_type(cls, value: Any) -> bool:
        """检查值是否是TOML支持的类型"""
        return (
            isinstance(value, (str, int, float, bool)) or
            value is None or
            isinstance(value, list) or
            isinstance(value, dict)
        )
    
    @classmethod
    def _merge_config(cls, current: Dict[str, Any], update: Dict[str, Any]) -> Dict[str, Any]:
        """合并配置数据（深度合并）"""
        merged = current.copy()
        
        for key, value in update.items():
            if value is not None:
                if (
                    isinstance(value, dict) and
                    key in merged and
                    isinstance(merged[key], dict)
                ):
                    # 递归合并嵌套对象
                    merged[key] = cls._merge_config(merged[key], value)
                else:
                    # 直接替换值
                    merged[key] = value
        
        return merged
    
    @classmethod
    def _update_tomlkit_doc(cls, doc: tomlkit.TOMLDocument, config: Dict[str, Any]) -> None:
        """使用tomlkit更新文档，保留注释和格式"""
        for key, value in config.items():
            if isinstance(value, dict):
                # 处理嵌套对象
                if key not in doc:
                    doc[key] = tomlkit.table()
                
                if isinstance(doc[key], tomlkit.items.Table):
                    cls._update_tomlkit_table(doc[key], value)
                else:
                    doc[key] = value
            else:
                # 直接设置值
                doc[key] = value
    
    @classmethod
    def _update_tomlkit_table(cls, table: tomlkit.items.Table, config: Dict[str, Any]) -> None:
        """更新tomlkit表格"""
        for key, value in config.items():
            if isinstance(value, dict):
                if key not in table:
                    table[key] = tomlkit.table()
                
                if isinstance(table[key], tomlkit.items.Table):
                    cls._update_tomlkit_table(table[key], value)
                else:
                    table[key] = value
            else:
                table[key] = value
    
    @classmethod
    async def _create_backup(cls, config_path: str) -> None:
        """创建配置文件备份"""
        try:
            main_root = path_cache_manager.get_main_root()
            if not main_root:
                raise RuntimeError('无法获取麦麦根目录路径')
            
            # 创建备份目录结构：config/LauncherConfigBak/main/
            backup_dir = os.path.join(main_root, 'config', 'LauncherConfigBak', 'main')
            os.makedirs(backup_dir, exist_ok=True)
            
            # 生成带时间戳的备份文件名
            timestamp = datetime.now().isoformat().replace(':', '-').replace('.', '-')
            config_filename = os.path.basename(config_path)
            backup_filename = f'{config_filename}.backup.{timestamp}'
            backup_path = os.path.join(backup_dir, backup_filename)
            
            # 复制配置文件到备份目录
            shutil.copy2(config_path, backup_path)
            logger.info(f'主程序配置备份已创建: {backup_path}')
            
        except Exception as error:
            logger.warning(f'创建配置文件备份失败: {error}')
            # 备份失败不应该阻止配置更新，只记录警告
